java/rocksjni/portal.h

```diff
@@ -2082,6 +2082,85 @@ class JniUtil {
 
       return jpointers;
     }
+
+    /*
+     * Helper for operations on a key and value
+     * for example WriteBatch->Put
+     *
+     * TODO(AR) could be extended to cover returning rocksdb::Status
+     * from `op` and used for RocksDB->Put etc.
+     */
+    static void kv_op_direct(
+        std::function<void(rocksdb::Slice&, rocksdb::Slice&)> op, JNIEnv* env,
+        jobject jkey, jint jkey_off, jint jkey_len, jobject jval, jint jval_off,
+        jint jval_len) {
+      char* key = reinterpret_cast<char*>(env->GetDirectBufferAddress(jkey));
+      if (key == nullptr ||
+          env->GetDirectBufferCapacity(jkey) < (jkey_off + jkey_len)) {
+        rocksdb::RocksDBExceptionJni::ThrowNew(env, "Invalid key argument");
```

Comment:
You are right. That is why I assert this everywhere in java like this:
  public int get(final ReadOptions opt, final ByteBuffer key, final ByteBuffer value)
    throws RocksDBException {
    assert key.isDirect() && value.isDirect();

Native functions are protected thus should not be possible to call it other way. Do you want to make it more defensive? Like check not using assert, but always? I was just following what is used on many places (asserts).

Reply:
Hmm I was thinking more that the message is perhaps misleading, as it could just be that their JVM does not implement the requisite NIO support in JNI. We could just make the error message more generic.

Reply:
OK I will try to come up with something more descriptive.

Reply:
I had to rebase, because of conflict, but message is update in latest change